packages/babel-plugin-proposal-partial-application/src/index.js

```diff
@@ -0,0 +1,279 @@
+   * and removes ArgumentPlaceholders type with Identifier
+   * and gives them a uniques name.
+   * @param {Object} node
+   * @param {Object} scope
+   * @returns {Array<Expression>} the arguments
+   */
+  function unwrapAllArguments(node, scope) {
+    const clone = t.cloneNode(node);
+
+    return clone.arguments.map(arg => {
+      if (arg.type === "ArgumentPlaceholder") {
+        return Object.assign({}, arg, {
+          type: "Identifier",
+          name: scope.generateUid("_argPlaceholder"),
+        });
+      }
+
+      return arg;
+    });
+  }
+
+  /**
+   * Makes an array of declarator for our VariableDeclaration
+   * @param {Array<Expression>} inits
+   * @param {Object} scope
+   */
+  function argsToVarDeclarator(inits, scope) {
+    return inits.map(expr => t.variableDeclarator(newParamLVal(scope), expr));
+  }
+
+  /**
+   * It replaces the values of non-placeholder args in allArgs
+   * @param {Array<Declarator>} nonPlaceholderDecl that has no placeholder in them
+   * @param {Array<Arguments>} args
+   */
+  function mapNonPlaceholderToLVal(nonPlaceholderDecl, allArgsList) {
+    const clone = Array.from(allArgsList);
+    clone.map(cl => {
+      nonPlaceholderDecl.forEach(dec => {
+        if (dec.init.type === cl.type) {
+          if (!!cl.value && cl.value === dec.init.value) {
+            cl.value = dec.id.name;
+          } else if (!!cl.name && cl.name === dec.init.name) {
+            cl.name = dec.id.name;
+          }
+        }
+      });
+    });
+    return clone;
+  }
+
+  /**
+   * Takes the full list of arguments and extracts placeholders from it
+   * @param {Array<Argument>} allArgsList full list of arguments
+   * @returns {Array<ArgumentPlaceholder>} cloneList
+   */
+  function placeholderLVal(allArgsList) {
+    let cloneList = [];
+    allArgsList.forEach(item => {
+      if (item.name && item.name.includes("_argPlaceholder")) {
+        cloneList = cloneList.concat(item);
+      }
+    });
+    return cloneList;
+  }
+
+  return {
+    name: "proposal-partial-application",
+    inherits: syntaxPartialApplication,
+
+    visitor: {
+      CallExpression(path) {
+        if (!hasArgumentPlaceholder(path.node)) {
+          return;
+        }
+        const { node, scope } = path;
+        const functionLVal = newFuncLVal(scope);
+        const receiverLVal = newReceiverLVal(scope);
+        const nonPlaceholderArgs = unwrapArguments(node);
+        const nonPlaceholderDecl = argsToVarDeclarator(
+          nonPlaceholderArgs,
+          scope,
+        );
+        const allArgs = unwrapAllArguments(node, scope);
+        const finalArgs = mapNonPlaceholderToLVal(nonPlaceholderDecl, allArgs);
+        const placeholderVals = placeholderLVal(allArgs);
+
+        if (node.callee.type === "MemberExpression") {
+          const data = [
+            t.variableDeclaration("const", [
+              t.variableDeclarator(
+                receiverLVal,
+                t.identifier(receiverRVal(node)),
```

Comment:
I don't understand what `receiverRVal` (which, btw, doesn't generate a valid identifier since the result can contain `.`) is needed for: isn't using `node.callee.object` (and `node.callee.property` for `functionLVal`) enough?

Reply:
Also, it doesn't work:

```js
a.b.fn().c["d"].e(?);

// ->

(() => {
  const _receiver = a.b.fn.c.undefined;
  const _func = a.b.fn.c.undefined.e;
  return _argPlaceholder => _func.call(_receiver, _argPlaceholder);
})();
```

Reply:
Oh, and instead of using `newFuncLVal`/`newReceiverLVal` you could use `path.scope.generateUidIdentifierBasedOnNode(node.callee)` and `path.scope.generateUidIdentifierBasedOnNode(node.callee.object)` to generate  names that make it easier to understand where they come from.
```js
object.get().fn(?);

// ->

(() => {
  const _object$get = object.get;
  const _object$get$fn = object.get.fn;
  return _argPlaceholder => _object$get$fn.call(_object$get, _argPlaceholder);
})();
```

Reply:
Oh, and the previous example should be something like this, otuherwise we might execute some accessors twice:
```js
(() => {
  const _object$get = object.get;
  const _object$get$fn = _object$get.fn;
  return _argPlaceholder => _object$get$fn.call(_object$get, _argPlaceholder);
})();
```